### DISCUSSION TITLE
Can I do order annotations at the module level?

### DISCUSSION THREAD

**Discussion by red8888** [pytest-dev/pytest-order#124] [Q&A] [answered]
All I need is for one particular module to run last is this possible?

I didn't see this anywhere in the docs. Want I would want is to be able to set the order at the top of the pytest file. For example I'd just set `@pytest.mark.order(99)` at the top of a module so it always runs last.

**Comment by mrbean-bremen** [comment]
This is not directly possible. A marker has always  to be set to a test or test class, and at least to my knowledge there is no easy way to set a marker to a whole module.

There is a workaround, however.
First, you use to use `--order-group-scope=module` to make sure that tests are not ordered between modules. Than you can set the order marker `@pytest.mark.order(-1)` or `@pytest.mark.order("last")`  to the _last_ test in that module.

That ensures that the tests inside the module are not reordered, as they are already ordered correctly, and the modules are afterwards ordered, so that the module with the marker is ordered last:

_test_one.py_
```py
def test_1():
    pass

def test_2():
    pass

def test_3():
    pass
```
_test_two.py_
```py
def test_1():
    pass

def test_2():
    pass

def test_3():
    pass
```
_test_last.py_
```py
import pytest

def test_1():
    pass

def test_2():
    pass

@pytest.mark.order(-1)
def test_3():
    pass
```
```
$ python -m pytest -v --order-group-scope=module
```
```
...
test_one.py::test1 PASSED     
test_one.py::test2 PASSED     
test_one.py::test3 PASSED     
test_two.py::test1 PASSED     
test_two.py::test2 PASSED     
test_two.py::test3 PASSED     
test_last.py::test1 PASSED    
test_last.py::test2 PASSED    
test_last.py::test3 PASSED    
```
(without that, `test_last` would run first due to its name).

**Comment by shaunc** [comment]
I think you can use a global "pytestmark" variable?

```py
pytestmark = [pytest.mark.order(1)]
```

cf https://docs.pytest.org/en/stable/example/markers.html#marking-whole-classes-or-modules

**Reply by mrbean-bremen** [reply, marked as answer]
Thanks, good point! I hadn't thought about this at the time.

What this does is essentially putting the marker on all tests in the module, and as the order is preserved for similar markers this will also work. You can change *test_last.py* to:
```py
import pytest

pytestmark = pytest.mark.order(-1)

def test_1():
    pass

def test_2():
    pass

def test_3():
    pass
```
so there is no need to put the marker onto the last test.

Note that you don't need to use `--order-group-scope=module` in caee you only use order markers with `pytestmark` (e.g. you are only ordering the modules themselves).

(updated to correct the need of `--order-group-scope=module`)
